In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
df_prods= pd.read_csv(r"C:\Users\abdul\Career-Foundry-Assignment4-Instacart-Basket-Analysis\02_Data\Original-Data\products.csv")

In [4]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [5]:
#1.If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.


In [6]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The max Prices = 99999.00 which is not correct.
Remaining every thing is correct. the corresponding Row for this price is displayed below.

In [7]:
df_prods[df_prods.prices==99999.000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [8]:
#2.Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
#Tip: Keep an eye on min and max values!


In [9]:
df_ords=pd.read_csv(r"C:\Users\abdul\Career-Foundry-Assignment4-Instacart-Basket-Analysis\02_Data\Prepared-Data\orders_wrangled.csv")

In [10]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   Unnamed: 0                    int64  
 1   orders_identification_number  int64  
 2   user_id                       int64  
 3   eval_set                      object 
 4   order_number                  int64  
 5   order_dow                     int64  
 6   order_hour_of_day             int64  
 7   days_since_prior_order        float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


In [11]:
df_ords.describe()

Unnamed: 0  orders_identification_number       user_id  order_number  \
count  3.421083e+06                  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06                  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05                  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00                  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05                  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06                  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06                  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06                  3.421083e+06  2.062090e+05  1.000000e+02   

          order_dow  order_hour_of_day  days_since_prior_order  
count  3.421083e+06       3.421083e+06            3.214874e+06  
mean   2.776219e+00       1.345202e+01            1.111484e+01  
std    2.046829e+00       4.226088e+00            9.206737e+00  
min    0.000000e+00       0.000000e+00            0.000000e+00  
25%    1.000000e+00       1.000000e+01            4.000000e+00  
50%    3.000000e+00       1.300000e+01            7.000000e+00  
75%    5.000000e+00       1.600000e+01            1.500000e+01  
max    6.000000e+00       2.300000e+01            3.000000e+01

#### In the "order_dow" column, max= 6  which is  less than the days of a week of 7.The min value = 0
#### In the "order_hour_of_day" column, the max= 23 which is less than the hours of a day of 24. The min value= 0.
#### IN the "days_since_prior_order", the min value = 0 which is not negative. 



In [12]:
#3.Check for mixed-type data in your df_ords dataframe.

In [13]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, 'mixed')
  else: print(col, 'consistent')

Unnamed: 0 consistent
orders_identification_number consistent
user_id consistent
eval_set consistent
order_number consistent
order_dow consistent
order_hour_of_day consistent
days_since_prior_order consistent


In [14]:
#4.If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

All the coloumns are single data type as shown above.

In [15]:
#5.Run a check for missing values in your df_ords dataframe.
#In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [16]:
df_ords.isnull().sum()

Unnamed: 0                           0
orders_identification_number         0
user_id                              0
eval_set                             0
order_number                         0
order_dow                            0
order_hour_of_day                    0
days_since_prior_order          206209
dtype: int64

In the column: 'days_since_prior_order' there are 206209 missing values are there. 

In [17]:
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [18]:
df_nan.head(20)

Unnamed: 0  orders_identification_number  user_id eval_set  order_number  \
0             0                       2539329        1    prior             1   
11           11                       2168274        2    prior             1   
26           26                       1374495        3    prior             1   
39           39                       3343014        4    prior             1   
45           45                       2717275        5    prior             1   
50           50                       2086598        6    prior             1   
54           54                       2565571        7    prior             1   
75           75                        600894        8    prior             1   
79           79                        280530        9    prior             1   
83           83                       1224907       10    prior             1   
89           89                       2411567       11    prior             1   
97           97                       1483446       12    prior             1   
103         103                       2618231       13    prior             1   
116         116                       2906263       14    prior             1   
130         130                       2514324       15    prior             1   
153         153                       1273676       16    prior             1   
160         160                       1737705       17    prior             1   
201         201                       2780889       18    prior             1   
208         208                       1321974       19    prior             1   
218         218                        947722       20    prior             1   

     order_dow  order_hour_of_day  days_since_prior_order  
0            2                  8                     NaN  
11           2                 11                     NaN  
26           1                 14                     NaN  
39           6                 11                     NaN  
45           3                 12                     NaN  
50           5                 18                     NaN  
54           3                  9                     NaN  
75           6                  0                     NaN  
79           1                 17                     NaN  
83           2                 14                     NaN  
89           0                 11                     NaN  
97           1                 12                     NaN  
103          6                 12                     NaN  
116          2                 18                     NaN  
130          5                 13                     NaN  
153          2                  8                     NaN  
160          2                 13                     NaN  
201          6                 18                     NaN  
208          6                 13                     NaN  
218          1                 10                     NaN

In the column: 'days_since_prior_order' there are 206209 missing values are there.
It is noted that in every 'user_id'starting 1,2,3... there is a missing value observed.

In [19]:
#6.Address the missing values using an appropriate method.
#In a markdown cell, explain why you used your method of choice.

In [20]:
df_ords.shape

(3421083, 8)

In [21]:
df_ords['days_since_prior_order'].value_counts()

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [22]:
df_ords['days_since_prior_order'].mean()

11.114836226863012

In [23]:
df_ords['days_since_prior_order'].median()

7.0

In [24]:
df_ords['days_since_prior_order'].fillna(7, inplace=True)

Explanation: Here the 30 is the most frequent value which influences the mean. the median is 7. 
The 7 is the second highest frequent value. if i put all NaN values then the analysis wont run. 
So i used the median value of 7 to replace the NaN values.

In [25]:
df_ords.isnull().sum()

Unnamed: 0                      0
orders_identification_number    0
user_id                         0
eval_set                        0
order_number                    0
order_dow                       0
order_hour_of_day               0
days_since_prior_order          0
dtype: int64

In [26]:
#7.Run a check for duplicate values in your df_ords data.
#In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [27]:
df_dups=df_ords[df_ords.duplicated()]

In [28]:
df_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, orders_identification_number, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [29]:
df_dups.size

0

In [30]:
#8.Address the duplicates using an appropriate method.
#In a markdown cell, explain why you used your method of choice.

There are no duplicates.

In [31]:
#9.Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [32]:
path=r"C:\Users\abdul\Career-Foundry-Assignment4-Instacart-Basket-Analysis"

In [33]:
df_ords.to_csv(os.path.join(path,'02_Data', 'Prepared-Data','orders_Cleaned.csv'))

In [34]:
#10.Save your Jupyter notebook and submit it here for your tutor to review.

Done